parsing/HACKING.adoc: explain more precisely how to promote a modified parser

shindere opened the pull request:
When the intention is to modify OCaml's parser itself, it is important to
use the same version of Menhir to generate the modified parser than the
one used to generate the original one.

This PR explains this and documents how to do so.

It has been suggested to the author of Menrhi, @fpottier, to make it
easier for users to find out which version of Menhir has been used to
produce a given parser.

At the moment, the `grep` command is the only way to find this version
information in `menhirLib`.

@fpottier seems to be willing to also include the version of menhir
and the command-line used to generate the parser as comments in the
generated parser files.

Personnally, I'd find it useful to also have at least the version
information available as an OCaml value exported by the generated parser
module (rather than just a comment), at least as a string.

Cc @gasche

shindere commented:
FWIW I came across the issue and was bitten while working on #11501.

Octachron commented:
I disagree that it is important to keep the same version of Menhir when updating the parser? The generated parser is a build artifact committed in the tree to avoid an external dependency. The source file is the mly file and is independent of the menhir version. Consequently, I think it is fine to increase the Menhir version when updating the parser.

shindere commented:
Florian Angeletti (2022/08/23 02:16 -0700):
> I disagree that it is important to keep the same version of Menhir
> when updating the parser? The generated parser is a build artifact
> committed in the tree to avoid an external dependency. The source file
> is the mly file and is independent of the menhir version.
> Consequently, I think it is fine to increase the Menhir version when
> updating the parser.

It seems @fpottier disagrees on this. Keeping the two kind of changes
separate is his recommendation, in a private e-mail so I'll let him
explain his position here.

Independently, it kind of makes sense to me to keep the two changes
separate from each other. It makes reverts easier, if necessary. I mean,
if a commit is not atomic and can be broken in two smaller bits, I can't
see anything wrong in doing so.

gasche commented on parsing/HACKING.adoc:
```diff
@@ -25,6 +25,52 @@ target of the root Makefile to rebuild the compiler parser. See
 Menhir-related targets and their use.
 
+In addition, if your purpose is to modify the parser, it is recommended that
+you make sure to use the same version of Menhir to generate the
```
"In addition, if your purpose if to modify the parser," is not so useful here, given that this section precisely explains how to modify the parser, so it should be clear from the context.

gasche commented on parsing/HACKING.adoc:
```diff
@@ -25,6 +25,52 @@ target of the root Makefile to rebuild the compiler parser. See
+Here, Menhir's version is `20210419` and it is thus this version
+of Menhir you need to use to re-generate the parser, to make sure the only
+modifications introduced in `boot/menhir` are those derived from
+your changes in `parser.mly`.
```
Note: I think it should always be okay to upgrade Menhirlib to a more recent version of Menhir, and never okay to downgrade to an older version. So the recommendation could be that if you have a more recent version of Menhir than was used in the compiler, you can have a (first, separate) commit to update MenhirLib, and then use your version.

gasche reviewed (commented):
I like the idea and many details are useful and clearly explained. But I would give different recommendations: instead of "try to use the same version by default", I would say "it's fine to upgrade to a more recent version if that is what you have on your system". This different policy results in up-to-date versions of Menhir being used, which makes it less likely than new contributors would find a mismatching version in the compiler tree.

gasche commented:
(I hadn't seen @Octachron's reply and I broadly agree. Personally I have no strong opinion on whether the version-upgrade should be in the same commit or a separate commit, either are fine.)

shindere commented:
Thanks a lot for your feedback, @gasche.

@fpottier also confirmed that it should not hurt to use a more version
of Menhir so I ammended the PR accordingly.

gasche commented on parsing/HACKING.adoc:
```diff
@@ -25,6 +25,69 @@ target of the root Makefile to rebuild the compiler parser. See
+If the version of Menhir you are using is more recent than the one
+used to generate the files under `boot/menhir`, then it is
+recommended that you do so in a dedicated commit which contains only
+this change and does not otherwise alter the parser.
```
In this paragraph it is not clear what "do so" means. I think the edits made the whole thing less clear -- it's hard to keep overall consistency when making local changes.

I would give the following key informations, in this order:
- if your menhir version is different from the one used for the last promotion, `promote-menhir` will change the versioned copy of menhirLib
  (you can use  ¨git status" to see this)
- we only accept these changes if your menhir version is *more recent* than the one used by the compiler
  (and we prefer if it comes as a separate commit before your main change)
- here is how to check for the version used in the compiler sources
- here is how to check for your version, get opam-available versions, install a specific version

shindere committed:
parsing/HACKING.adoc: explain more precisely how to promote a new parser

When the intention is to modify the grammar, it is important to make
sure that the version of Menhir used to generate the updated parser is
at least as new as the one used to generate the current parser.

This commit explains this and documents how to proceed.

shindere commented:
Many thanks for your feedback, @gasche.

I fully agree with you that the local edits made the text less clear than
it initially was. I have thus rewritten it from scratch and believe it is now
even clearer than it was at the beginning.

Let's see what you think about this version. I am unsure about the
formatting and rendering of `.adoc` files so please do not hesitate to
provide feedback especially at this typographic level.
